test/Microsoft.ML.Benchmarks/TrainPredictionBench.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Running;
+using Microsoft.ML.Models;
+using Microsoft.ML.Runtime.Api;
+using Microsoft.ML.Trainers;
+using Microsoft.ML.Transforms;
+
+namespace Microsoft.ML.Benchmarks
+{
+    public class TrainPredictionBench
+    {
+        internal static ClassificationMetrics s_metrics;
+        static PredictionModel<IrisData, IrisPrediction> s_trainedModel;
+
+        [GlobalCleanup]
+        public void Accuracy()
+        {
+            var dataPath = Program.GetDataPath("iris.txt");
+            var testData = new TextLoader<IrisData>(dataPath, useHeader: true, separator: "tab");
+            var evaluator = new ClassificationEvaluator();
+            s_metrics = evaluator.Evaluate(s_trainedModel, testData);
+        }
+
+        [Benchmark]
+        public void Iris()
+        {
+            var dataPath = Program.GetDataPath("iris.txt");
+            var pipeline = new LearningPipeline();
+
+            pipeline.Add(new TextLoader<IrisData>(dataPath, useHeader: true, separator: "tab"));
+            pipeline.Add(new ColumnConcatenator(outputColumn: "Features",
+                "SepalLength", "SepalWidth", "PetalLength", "PetalWidth"));
+
+            pipeline.Add(new StochasticDualCoordinateAscentClassifier());
+
+            PredictionModel<IrisData, IrisPrediction> model = pipeline.Train<IrisData, IrisPrediction>();
+
+            IrisPrediction prediction = model.Predict(new IrisData()
+            {
+                SepalLength = 3.3f,
+                SepalWidth = 1.6f,
+                PetalLength = 0.2f,
+                PetalWidth = 5.1f,
+            });
+
+            prediction = model.Predict(new IrisData()
```

Comment:
>Predict [](start = 31, length = 7)

There are two dimensions on measuring predictions:
1. Measiuring pipeline creation time
2. Measuring each individual prediction once pipeline is established.
Often the two differ by 100x.  Without this separation the results are very misleading
In high perf systems a great care must be taken to minimize initialization of predictors.

Reply:
Is it in addition to splitting training and prediction? I created two models (one cold and one hot) and ran the benchmark. There was no difference in the results.

Reply:
Quick question  - can you compare times of running 1,2 and 5 predictions. Are they linear (1,2 5x)?   I believe the time is taken when the enumeration is initialized (per predict call)
If the times for tests above are nearly equal, then it means that most time is taken with pipeline initialization.

---
In reply to: [187092994](https://github.com/dotnet/machinelearning/pull/62#discussion_r187092994) [](ancestors = 187092994)

Reply:
They scale nicely:

``` ini

BenchmarkDotNet=v0.10.14, OS=Windows 10.0.16299.431 (1709/FallCreatorsUpdate/Redstone3)
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
Frequency=3328124 Hz, Resolution=300.4696 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008662
  [Host] : .NET Core 2.0.7 (CoreCLR 4.6.26328.01, CoreFX 4.6.26403.03), 64bit RyuJIT

Toolchain=InProcessToolchain  LaunchCount=1  TargetCount=3  
WarmupCount=3  

```
|      Method | NumberOfPredictions |       Mean |       Error |     StdDev | AccuracyMacro |     Gen 0 |    Gen 1 |   Gen 2 |   Allocated |
|------------ |-------------------- |-----------:|------------:|-----------:|--------------:|----------:|---------:|--------:|------------:|
| **PredictIris** |                   **1** |   **1.011 ms** |   **0.4383 ms** |  **0.0248 ms** |          **0.98** |   **27.3438** |  **13.6719** |  **1.9531** |    **88.15 KB** |
| **PredictIris** |                   **2** |   **2.145 ms** |   **3.0685 ms** |  **0.1734 ms** |          **0.98** |   **54.6875** |  **27.3438** |  **3.9063** |   **176.31 KB** |
| **PredictIris** |                   **5** |   **5.326 ms** |   **8.2691 ms** |  **0.4672 ms** |          **0.98** |  **132.8125** |  **70.3125** | **15.6250** |   **440.45 KB** |

Reply:
Cool.  Is it possible to add a test that takes an enumeration of 1, 2, and 5 data points for each call of Predict?  I"d expect it to be a lot faster.

---
In reply to: [187421073](https://github.com/dotnet/machinelearning/pull/62#discussion_r187421073) [](ancestors = 187421073)

Reply:
Need to run tests with several examples for a single Predict call, not several calls to predict. They are not the same.

---
In reply to: [187246794](https://github.com/dotnet/machinelearning/pull/62#discussion_r187246794) [](ancestors = 187246794,187092994)